csrc/scheduler/matmul_utils.cpp

```diff
@@ -314,144 +314,302 @@ std::string isMatmulFusionDefinitionSupported(
+  // even if it has the same role as the innermost dimension in the canonical
+  // ordering.
+  int64_t operandVectorization(TensorView* tv) {
+    // Check data pointer alignment
+    int64_t vec_size = ptrAndDTypeVec(tv);
+    if (vec_size == 1l) {
+      return vec_size;
     }
+
+    // Find the inner-most non-batch role for this tensor, and collect all
+    // ValGroups in that role, in the canonical ordering.
+    std::optional<MatmulDomain> vec_dim_role = std::nullopt;
+    for (int64_t i = (int64_t)(tv->getMaybeAllocationDomain().size()) - 1;
+         i >= 0;
+         --i) {
+      IterDomain* id = tv->getMaybeAllocationDomain()[i];
+      if (id->isDeviceDim() || id->isReduction() || id->isBroadcast()) {
+        continue;
+      }
+
+      ValGroup g = exact_graph_.toGroup(id);
+      MatmulDomain dim_role = dimRole(g);
+      if (dim_role == MatmulDomain::Batch) {
+        // We cannot vectorize in batch dimensions
+        break;
+      }
+      if (!vec_dim_role.has_value()) {
+        vec_dim_role = dim_role;
+      }
+    }
+    if (!vec_dim_role.has_value()) {
+      // Didn't find any dimensions to vectorize
+      return 1l;
+    }
+
+    // Extract dims with this role in the canonical ordering
+    std::vector<ValGroup> ordered_inner_dims;
+    for (const ValGroup& og : dim_ordering_) {
+      if (dimRole(og) == vec_dim_role.value()) {
+        ordered_inner_dims.push_back(og);
+      }
+    }
+
+    return std::min(vec_size, innerDimsVectorization(tv, ordered_inner_dims));
   }
-  return supported_vec_size;
+
+  int64_t epilogueVectorization() {
+    // This is a vector of non-K dimensions sorted from inner to outer
+    std::vector<ValGroup> inner_nonk_dims;
+    std::optional<MatmulDomain> inner_nonk_role = std::nullopt;
+    for (auto g_it = dim_ordering_.rbegin(); g_it != dim_ordering_.rend();
+         ++g_it) {
+      const ValGroup& g = *g_it;
+
+      MatmulDomain dim_role = dimRole(g);
+      if (dim_role == MatmulDomain::K) {
+        // Skip K dims since they won't appear in epilogue loop nest
+        continue;
+      }
+      if (!inner_nonk_role.has_value()) {
+        inner_nonk_role = dim_role;
+      }
+      if (dim_role != inner_nonk_role.value()) {
+        break;
+      }
+      inner_nonk_dims.push_back(g);
+    }
+
+    if (!inner_nonk_role.has_value() ||
+        inner_nonk_role.value() == MatmulDomain::Batch) {
+      // If the innermost non-K dimension is a batch dimension, then we cannot
+      // vectorize the outputs since we parallelize batch dimensions across the
+      // grid.
+      return 1l;
+    }
+
+    // Match the innermost dimensions above to contiguous innermost dims in tv
+    // from inner to outer. Determine supported vectorization based on product
+    // of matching sizes along with all outer strides.
+    const auto innerMostVec = [&](TensorView* tv) {
+      return std::min(
+          ptrAndDTypeVec(tv), innerDimsVectorization(tv, inner_nonk_dims));
```

Comment:
Should we exclude batch from the argument of `innerDimsVectorization`?

Reply:
Also, should we check strides?

Reply:
> Should we exclude batch from the argument of `innerDimsVectorization`?

Just above this we return 1 if `inner_nonk_role` is Batch. The `inner_nonk_dims` all have this role, and since it's not K, then all `inner_nonk_dims` must be M or N for this call.

> Also, should we check strides?

Strides are checked in `innerDimsVectorization` but not beyond dimensions marked contiguous in the `TensorView`. See my reply above.

Reply:
> Just above this we return 1 if `inner_nonk_role` is Batch. The `inner_nonk_dims` all have this role, and since it's not K, then all `inner_nonk_dims` must be M or N for this call.

I see, somehow I missed that the `inner_nonk_dims` contain dims of only a single role.


> Strides are checked in `innerDimsVectorization` but not beyond dimensions marked contiguous in the `TensorView`. See my reply above.

See my reply above. I believe we need to use these strides to calculate vector size.

Reply:
> See my reply above. I believe we need to use these strides to calculate vector size.

Oh OK I see what you're saying. Yes, `innerDimsVectorization` does check the outer strides. How it works is we look through the provided dimensions and find the most of those that we can vectorize. We look at the combined size to compute an initial vectorization. Then we start at that position and look at all outer strides starting at that location, which might potentially reduce the vectorization width if they are not multiples of the vectorization.

Reply:
Oh, I see. I missed that. Find it now.